app/renderer/util.js

```diff
@@ -114,69 +120,124 @@ function isXpathUnique(xpath, doc, domNode) {
   return [true];
 }
 
+/**
+ * Given an xml doc and a current dom node, try to find a robust xpath selector qualified by
+ * key attributes, which is unique in the document (or unique plus index).
+ *
+ * @param {string} xpath
+ * @param {DOMDocument} doc
+ * @param {Array<string>|Array<[string, string]>} attrs - a list of attributes to consider, or
+ * a list of pairs of attributes to consider in conjunction
+ * @param {boolean} isPairs - whether we are looking for single attributes or a conjunction of
+ * attributes
+ *
+ * @returns {[string?, boolean?]} tuple consisting of (1) the xpath selector discovered, and (2)
```

Comment:
actually question marks mean the same as `string|null`. For undefined case above `string|undefined` should be used instead